changefeedccl: fix split_column_families changefeeds

KeithCh opened the pull request:
This fixes the following when creating changefeeds 
for tables that have multiple columns and with the 
option split_column_families, into sinks that have topics:

When using a Kafka V1 sink, we would always 
attempt to create a nonsense topic even if the
Kafka broker setting has
auto.create.topics.enable=false. This would cause
the changefeed to fail.

Release note(bug fix):
Remove unncessary Kafka topic creation that would
could cause changefeed start up to fail when using 
changefeed.new_kafka_sink_enabled=false.

Fixes: https://github.com/cockroachdb/cockroach/issues/128973
Epic: CRDB-41784

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146476)
<!-- Reviewable:end -->

KeithCh commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -5126,6 +5126,70 @@ func TestChangefeedOutputTopics(t *testing.T) {
+				}
+
+				actual = "(no notice)"
+				f := func() cdctest.TestFeedFactory {
```
Context for why I didn't use `cdcTest` 
https://github.com/cockroachdb/cockroach/pull/144859#discussion_r2082319959

asg0451 commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -264,9 +264,8 @@ func (s *kafkaSink) Dial() error {
-		// Now that we do not fetch metadata for all topics by default, we try
-		// RefreshMetadata manually to check for any connection error.
+	if err = client.RefreshMetadata(); err != nil {
+		// Try RefreshMetadata manually to check for any connection error.
```
i think we dont fetch all topics on purpose in case we don't have access to some, if the user has set that up

KeithCh replied:
Changed to a different method that makes an RPC call.

asg0451 commented on pkg/ccl/changefeedccl/changefeed_test.go:
why is the port necessary?

KeithCh replied:
You would get this error
`pq: dial tcp: address nope: missing port in address`

asg0451 commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -264,10 +267,16 @@ func (s *kafkaSink) Dial() error {
-		// RefreshMetadata manually to check for any connection error.
-		return errors.CombineErrors(err, client.Close())
+	// Make sure the broker can be reached.
+	broker := client.LeastLoadedBroker()
```
i think this will mostly just end up being the seed broker(s), but that's ok

asg0451 commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -264,10 +267,16 @@ func (s *kafkaSink) Dial() error {
-		return errors.CombineErrors(err, client.Close())
+	// Make sure the broker can be reached.
+	broker := client.LeastLoadedBroker()
+	if broker != nil {
```
when can this be nil? if it is nil should we error?

asg0451 commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -264,10 +267,16 @@ func (s *kafkaSink) Dial() error {
+	// Make sure the broker can be reached.
+	broker := client.LeastLoadedBroker()
+	if broker != nil {
+		req := &sarama.ApiVersionsRequest{}
```
the arguments for the ApiVersionsRequest are kinda sarama specific and i'd hate to have to replicate them.

instead i think you should do this (pseudocode):

```go
if err := broker.Open(s.kafkaCfg); err != nil && !errors.Is(err, sarama.AlreadyConnected) { .. }
if ok, err := broker.Connected(); !ok || err != nil { ... }
```

see https://pkg.go.dev/github.com/shopify/sarama#Broker.Open

KeithCh committed:
changefeedccl: fix split_column_families changefeeds

This fixes the following when creating changefeeds
for tables that have multiple columns and with the
option split_column_families, into sinks that have
topics:

When using a Kafka V1 sink, we would always
attempt to create a nonsense topic even if the
Kafka broker setting has
auto.create.topics.enable=false. This would cause
the changefeed to fail.

Release note(bug fix):
Remove unncessary Kafka topic creation that would
could cause changefeed start up to fail when using
changefeed.new_kafka_sink_enabled=false.

Fixes: #128973
Epic: CRDB-41784

KeithCh commented on pkg/ccl/changefeedccl/topic.go:
```diff
@@ -185,7 +185,7 @@ func (tn *TopicNamer) makeName(s changefeedbase.Target, td TopicDescriptor) (str
 	case jobspb.ChangefeedTargetSpecification_EACH_FAMILY:
 		if td == nil {
-			return tn.nameFromComponents(s.StatementTimeName, familyPlaceholder), nil
+			return tn.nameFromComponents(s.StatementTimeName) + string(tn.join) + familyPlaceholder, nil
```
This fixes the  `t.u007b_family_u007d` issue so that it'll show as `t.{family}` instead

rharding6373 commented:
[Team meeting]: Since this is an issue with kafka v1, we don't currently plan on backporting the fix.
